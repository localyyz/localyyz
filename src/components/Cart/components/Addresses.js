import React from "react";
import { Alert, View, StyleSheet, Text, TouchableOpacity } from "react-native";
import { Colours, Sizes, Styles } from "localyyz/constants";
import PropTypes from "prop-types";

// custom
import Address from "./Address";
import AddressForm from "./AddressForm";
import CartHeader from "./CartHeader";

// third party
import EntypoIcon from "react-native-vector-icons/Entypo";
import FontAwesomeIcon from "react-native-vector-icons/FontAwesome";
import {
  inject,
  observer,
  PropTypes as mobxPropTypes
} from "mobx-react/native";

@inject((props, stores) => ({
  addresses: (props.addressStore || stores.addressStore).addresses,
  fetch: () => (props.addressStore || stores.addressStore ).fetch(),
  remove: address => (props.cartStore || stores.cartStore).removeAddress({
      address: address
    }),
  update: address => (props.cartStore || stores.cartStore).updateAddress({
      address: address
    })
}))
@observer
export default class Addresses extends React.Component {
  static propTypes = {
    address: PropTypes.object.isRequired,
    title: PropTypes.string,

    // mobx injected
    addresses: mobxPropTypes.arrayOrObservableArray.isRequired,
    fetch: PropTypes.func.isRequired,
    update: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired
  };

  static defaultProps = {
    title: "",
    address: {}
  };

  constructor(props) {
    super(props);

    // initial states
    this.state = {
      // current address from prop
      currentAddress: props.address,
      // isSelecting opens address(es) selection list
      isSelecting: props.address.address,
      // isEditing opens editable address form
      isEditing: props.address.hasError || this.props.addresses.length === 0,
      // isComplete shows completed status
      isComplete:
        props.address && props.address.address && !props.address.hasError,
      // isShipping and isBilling defines type of address this component is
      // editing
      isShipping: props.address.isShipping,
      isBilling: props.address.isBilling
    };
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (nextProps.address.hasError) {
      this.setState({
        isSelecting: true,
        isEditing: true,
        isComplete: false
      });
    } else {
      this.setState({
        currentAddress: nextProps.address,
        // isSelecting opens address(es) selection list
        isSelecting: nextProps.address.address,
        // isEditing opens editable address form
        isEditing:
          nextProps.address.hasError || nextProps.addresses.length === 0,
        // isComplete shows completed status
        isComplete: !!nextProps.address.address
      });
    }
  }

  componentDidMount() {
    this.props.fetch();
  }

  toggleSelect = selecting => {
    // determine if selecting
    const isSelecting = selecting != null ? selecting : !this.state.isSelecting;

    this.setState({
      isSelecting: isSelecting,
      // empty addresses automatically toggle isEditing
      isEditing: isSelecting && this.props.addresses.length === 0
    });
  };

  onAddressUpdate = address => {
    this.props.update({
      ...address,
      isShipping: this.state.isShipping,
      isBilling: this.state.isBilling
    });

    this.setState({
      isEditing: false
    });
  };

  onAddressRemove = address => {
    Alert.alert("Remove this address?", null, [
      {
        text: "Cancel",
        style: "cancel"
      },
      {
        text: "Remove",
        onPress: () => {
          this.props.remove(address);
        }
      }
    ]);
  };

  get renderEditAddress() {
    return this.state.currentAddress.id ? (
      <View style={styles.addAddress}>
        <TouchableOpacity
          onPress={() => {
            this.setState({ isEditing: true });
          }}>
          <Text
            style={[
              Styles.Text,
              Styles.Terminal,
              Styles.Emphasized,
              Styles.Underlined,
              styles.addAddressLabel
            ]}>
            edit current address
          </Text>
        </TouchableOpacity>
      </View>
    ) : null;
  }

  get renderCreateAddress() {
    return this.props.addresses.length > 0 ? (
      <View style={styles.addAddress}>
        <TouchableOpacity
          onPress={() => {
            this.setState({ isEditing: true });
          }}>
          <Text
            style={[
              Styles.Text,
              Styles.Terminal,
              Styles.Emphasized,
              Styles.Underlined,
              styles.addAddressLabel
            ]}>
            add a new address
          </Text>
        </TouchableOpacity>
      </View>
    ) : null;
  }

  get renderAddressForm() {
    return (
      <AddressForm
        address={this.state.currentAddress}
        onSubmit={address => this.onAddressUpdate(address)}/>
    );
  }

  get renderAddresses() {
    return (
      <View style={styles.addresses}>
        {!this.state.isEditing ? (
          <View>
            {this.props.addresses.slice().map(address => (
              <Address
                key={`address-${address.id}`}
                address={address}
                buttonIcon="trash"
                onActionPress={this.onAddressRemove}
                onPress={() => {
                  this.setState({
                    currentAddress: address,
                    isEditing: true
                  });
                }}/>
            ))}
            {this.renderCreateAddress}
          </View>
        ) : (
          this.renderAddressForm
        )}
      </View>
    );
  }

  render() {
    return (
      <View style={styles.container}>
        <TouchableOpacity onPress={() => this.toggleSelect()}>
          <CartHeader
            ref="header"
            title={this.props.title}
            icon={
              this.state.isComplete ? (
                <FontAwesomeIcon
                  name="check-circle"
                  size={Sizes.IconButton / 2}
                  color={Colours.PositiveButton}
                  style={Styles.IconOffset}/>
              ) : (
                <EntypoIcon
                  name="dot-single"
                  size={Sizes.IconButton}
                  color={Colours.NegativeButton}/>
              )
            }>
            {this.state.isComplete
              ? this.state.currentAddress.shortAddress
              : "no address selected"}
          </CartHeader>
        </TouchableOpacity>
        {this.state.isSelecting ? this.renderAddresses : null}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {},

  addAddress: {
    padding: Sizes.InnerFrame,
    alignItems: "flex-end"
  }
});
