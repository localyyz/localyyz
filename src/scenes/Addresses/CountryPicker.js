import React from "react";
import {
  StyleSheet,
  FlatList,
  Text,
  View,
  TouchableHighlight
} from "react-native";

import { Styles, Sizes, Colours } from "~/src/constants";
import { deviceStore } from "~/src/stores";

import AddressStore from "~/src/stores/AddressStore";

export default class CountryPicker extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      countries: [],
      _options: [],
      _current: [],
      _page: 1,
      _limit: 10
    };

    // must be required props
    this.update = props.navigation.getParam("update");
  }

  componentDidMount = () => {
    const deviceCountryCode = deviceStore.getDeviceCountry();
    AddressStore.fetchCountries().then(resolved => {
      if (resolved.countries) {
        let countries = [];
        for (let country of resolved.countries) {
          let c = {
            value: country.alpha2Code,
            name: country.name,
            nativeName: country.nativeName,
            key: country.alpha2Code
          };

          if (c.key == deviceCountryCode) {
            countries.unshift({
              ...c,
              isSuggested: true,
              description: "Suggested country:"
            });
          } else {
            countries.push(c);
          }
        }

        this.setState({
          countries: countries,
          _options: countries.slice(0, this.state._limit)
        });
      }
    });
  };

  onEndReached = () => {
    let start = this.state._page * this.state._limit;
    let end = start + this.state._limit;

    if (start > this.state.countries.length) {
      return;
    }

    this.setState({
      _options: this.state._options.concat(
        this.state.countries.slice(start, end)
      ),
      _page: this.state._page + 1
    });
  };

  onSelect = value => {
    this.update(value);
  };

  renderItem = ({ item: country }) => {
    const label
      = country.name == country.nativeName ? country.name : country.nativeName;

    return (
      <TouchableHighlight
        style={{ width: Sizes.Width }}
        onPress={() =>
          this.onSelect({
            country: country.nativeName,
            countryCode: country.value
          })
        }
        underlayColor={Colours.Background}>
        <View
          style={[styles.row, country.isSuggested ? styles.suggestedRow : {}]}>
          {country.description ? (
            <Text style={styles.rowDescription}>{country.description}</Text>
          ) : null}
          <Text style={styles.rowText}>{label}</Text>
        </View>
      </TouchableHighlight>
    );
  };

  render() {
    return (
      <FlatList
        renderItem={this.renderItem}
        keyExtractor={o => o.key}
        scrollEventThrottle={16}
        maxToRenderPerBatch={10}
        initialNum={10}
        removeClippedSubviews={true}
        data={this.state._options}
        contentContainerStyle={styles.container}
        onEndReached={this.onEndReached}
        ItemSeparatorComponent={() => (
          <View
            style={{
              height: 1,
              borderBottomWidth: Sizes.Hairline,
              borderBottomColor: Colours.Border
            }}/>
        )}/>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colours.Foreground,
    paddingTop: Sizes.OuterFrame,
    paddingBottom: Sizes.IOSTabBar + Sizes.OuterFrame * 2
  },

  row: {
    paddingVertical: Sizes.InnerFrame,
    marginHorizontal: Sizes.InnerFrame,
    justifyContent: "center",
    height: Sizes.Height / 10
  },

  suggestedRow: {
    marginBottom: Sizes.OuterFrame
  },

  rowText: {
    ...Styles.Text,
    ...Styles.Emphasized
  },

  rowDescription: {
    ...Styles.SmallText,
    color: Colours.SubduedText,
    paddingBottom: 5
  }
});
