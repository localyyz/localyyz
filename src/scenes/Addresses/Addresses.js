import React from "react";
import {
  View,
  Text,
  StyleSheet,
  FlatList,
  TouchableHighlight
} from "react-native";

// third party
import { inject, observer } from "mobx-react/native";
import MaterialIcon from "react-native-vector-icons/MaterialIcons";

// custom
import { Colours, Styles, Sizes } from "localyyz/constants";
import { Forms } from "localyyz/components";

@inject(stores => ({
  addresses: stores.addressStore.addresses.slice(),
  fetch: stores.addressStore.fetch
}))
@observer
export default class Addresses extends React.Component {
  static navigationOptions = ({ navigation, navigationOptions }) => ({
    ...navigationOptions,
    title: "Saved Addresses",
    headerRight: (
      <MaterialIcon.Button
        color={"black"}
        backgroundColor={Colours.Transparent}
        onPress={() => navigation.navigate("AddressForm", { title: "Add New" })}
        underlayColor={Colours.Transparent}
        style={{
          justifyContent: "flex-end",
          paddingRight: Sizes.InnerFrame,
          alignItems: undefined
        }}
        iconStyle={{
          marginRight: undefined
        }}
        name="add"
        size={32}
        hitSlop={{
          top: Sizes.OuterFrame,
          bottom: Sizes.OuterFrame,
          left: Sizes.OuterFrame,
          right: Sizes.OuterFrame
        }}/>
    )
  });

  constructor(props) {
    super(props);
    this.settings = this.props.navigation.state.params || {};

    // bindings
    this.renderItem = this.renderItem.bind(this);
    this.onPress = this.onPress.bind(this);
    this.onSelect = this.onSelect.bind(this);
    this.addAddress = this.addAddress.bind(this);
    this.updateAddress = this.updateAddress.bind(this);
  }

  componentDidMount() {
    this.props.fetch();
  }

  renderItem({ item: address }) {
    return (
      <TouchableHighlight
        style={{ width: Sizes.Width }}
        onPress={() => this.onPress(address)}
        underlayColor={Colours.Background}>
        <View style={styles.row}>
          <Text style={styles.rowText}>{address.address}</Text>
          <View style={styles.arrow}>
            <MaterialIcon
              name="keyboard-arrow-right"
              size={Sizes.Text}
              color={Colours.SubduedText}/>
          </View>
        </View>
      </TouchableHighlight>
    );
  }

  onPress(address) {
    !this.onSelect(address) && this.updateAddress(address);
  }

  onSelect(address) {
    if (this.settings.onSelect) {
      this.settings.onSelect(address);

      // and out
      this.props.navigation.goBack(null);

      // help skip updateAddress on onPress
      return true;
    }
  }

  addAddress() {
    return this.props.navigation.navigate("AddressForm", {
      onSubmit: this.onSelect,
      title: "Add New"
    });
  }

  updateAddress(address) {
    return this.props.navigation.navigate("AddressForm", {
      shouldUpdate: true,
      address: address
    });
  }

  addButton = () => {
    return (
      <Forms.Button isEnabled onPress={() => this.addAddress()}>
        Add a new address
      </Forms.Button>
    );
  };

  empty = () => {
    return (
      <View style={styles.footer}>
        <Text style={styles.subtitle}>
          Looks like we don't have an address for you yet.
        </Text>
        {this.addButton()}
      </View>
    );
  };

  render() {
    return (
      <FlatList
        style={styles.container}
        contentContainerStyle={{
          paddingBottom: this.props.addresses.length
            ? Sizes.OuterFrame * 2 + Sizes.IOSTabBar
            : 0
        }}
        data={this.props.addresses}
        keyExtractor={address => `${address.id}`}
        renderItem={this.renderItem}
        ListEmptyComponent={this.empty}
        ItemSeparatorComponent={() => (
          <View
            style={{
              height: 1,
              borderBottomWidth: Sizes.Hairline,
              borderBottomColor: Colours.Border
            }}/>
        )}/>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colours.Foreground
  },

  row: {
    paddingVertical: Sizes.InnerFrame,
    marginHorizontal: Sizes.InnerFrame,
    justifyContent: "space-between",
    alignItems: "center",
    flexDirection: "row",
    height: Sizes.Height / 10
  },

  rowText: {
    ...Styles.Text,
    ...Styles.Emphasized
  },

  subtitle: {
    ...Styles.SectionSubtitle,
    textAlign: "center"
  },

  footer: {
    paddingTop: Sizes.OuterFrame,
    paddingHorizontal: Sizes.OuterFrame
  }
});
