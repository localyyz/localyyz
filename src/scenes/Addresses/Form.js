import React from "react";
import {
  Keyboard,
  KeyboardAvoidingView,
  SafeAreaView,
  View,
  StyleSheet,
  ScrollView
} from "react-native";

// third party
import { Provider, observer, inject } from "mobx-react/native";

// custom
import { Styles, Sizes, Colours } from "localyyz/constants";
import { Forms } from "localyyz/components";
import { UserAddress } from "localyyz/models";
import { navbarStore } from "~/src/stores";

import GoogleAddress from "./GoogleAddress";

// constants
const DEBUG = false;
const SERVER_FAILURE
  = "There was an internal server error and your request could not be processed";

@inject(stores => ({
  add: address => stores.addressStore.add(address),
  update: address => stores.addressStore.update(address),
  delete: address => stores.addressStore.remove(address.id)
}))
@observer
export default class AddressForm extends React.Component {
  static navigationOptions = ({ navigation, navigationOptions }) => ({
    ...navigationOptions,
    title: navigation.getParam("title", "Edit")
  });

  constructor(props) {
    super(props);
    this.settings
      = (this.props.navigation
        && this.props.navigation.state
        && this.props.navigation.state.params)
      || {};
    this.store = new Forms.Store(
      // internally used
      "id",

      // associated form fields
      {
        id: "firstName",
        validators: ["isRequired"]
      },
      { id: "lastName", validators: ["isRequired"] },
      { id: "street", validators: ["isRequired"] },
      "addressOpt",
      { id: "city", validators: ["isRequired"] },
      { id: "region" },
      { id: "postal", validators: ["isRequired"] },
      {
        id: "country",
        validators: ["isRequired"],
        options: {}
      },
      {
        id: "address",

        // mapping UserAddress fields to these fields on update
        mapping: {
          address: "street",
          city: "city",
          province: "region",
          zip: "postal"
        }
      }
    );
    this.state = { manual: false };
  }

  componentDidMount() {
    // merge in outside data if provided, if address was never touched,
    // allowing address component field syncing break it up into
    // all the associated fields
    if (this.settings.address) {
      for (let key in this.settings.address) {
        // NOTE: there are some weird mapping thing going on here, for
        // now just blindly support it
        let mappingKey = "";
        switch (key) {
          case "address":
            mappingKey = "street";
            break;
          case "province":
            mappingKey = "region";
            break;
          case "zip":
            mappingKey = "postal";
            break;
          default:
            mappingKey = key;
        }
        this.store.update(mappingKey, this.settings.address[key]);
      }
    }
  }

  submit = async () => {
    try {
      // attempt update or add
      let address = await (this.shouldUpdate
        ? this.props.update
        : this.props.add)(this.address);

      // bail prematurely due to server failure
      // TODO: call setError for field errors on status 422
      if (address.hasError) {
        throw SERVER_FAILURE;
      }

      // if callback given
      this.settings.onSubmit
        && this.settings.onSubmit(new UserAddress(address));

      // alternative to navigation
      if (this.settings.onComplete) {
        this.settings.onComplete();
      } else {
        this.props.navigation.goBack(null);
      }
    } catch (err) {
      DEBUG && console.log("AddressForm: Failed to set address.", err);
      navbarStore.notify(`${err} - Please try again later`, {
        backgroundColor: Colours.Fail
      });
    }
  };

  delete = async () => {
    try {
      let response = await this.props.delete(this.address);
      if (response.hasError) {
        throw SERVER_FAILURE;
      }

      // if callback given
      this.settings.onDelete && this.settings.onDelete();

      // alternative to navigation
      if (this.settings.onComplete) {
        this.settings.onComplete();
      } else {
        this.props.navigation.goBack();
      }
    } catch (err) {
      DEBUG && console.log("AddressForm: Failed to delete address.", err);
      navbarStore.notify(`${err} - Please try again later`, {
        backgroundColor: Colours.Fail
      });
    }
  };

  get shouldUpdate() {
    return !!this.address && !!this.address.id;
  }

  get submitButton() {
    return (
      this.settings.buttonLabel
      || (this.shouldUpdate ? "Update this address" : "Add this address")
    );
  }

  get deleteButton() {
    return this.settings.deleteButtonLabel || "Remove this address";
  }

  get title() {
    return this.settings.title || this.shouldUpdate
      ? "Update address details"
      : "Add a new address";
  }

  get address() {
    // these are basically reverse of the above merge object, should be in
    // sync with the mapping above
    return this.store.isComplete
      ? new UserAddress({
          id: this.store.data.id,
          address: this.store.data.street,
          city: this.store.data.city,
          province: this.store.data.region,
          zip: this.store.data.postal,
          country: this.store.data.country,
          firstName: this.store.data.firstName,
          lastName: this.store.data.lastName,
          addressOpt: this.store.data.addressOpt
        })
      : null;
  }

  pickCountry = () => {
    this.props.navigation.navigate("CountryPicker", {
      field: "country",
      title: "Select a Country",
      update: value => this.store.update("country", value)
    });
  };

  onSubmitEditing = () => {
    const nextField = this.store.nextField;
    if (nextField) {
      switch (nextField.id) {
        case "firstName":
          this._firstNameRef.focus();
          break;
        case "lastName":
          this._lastNameRef.focus();
          break;
        case "city":
          this._cityRef.focus();
          break;
        case "street":
          this._streetRef.focus();
          break;
        case "postal":
          this._postalRef.focus();
          break;
        case "country":
          this._countryRef.focus();
          break;
      }
    } else {
      Keyboard.dismiss();
    }
  };

  renderManualForm = () => {
    return (
      <KeyboardAvoidingView style={styles.container} behavior="padding" enabled>
        <ScrollView contentContainerStyle={styles.content}>
          <View style={Styles.Horizontal}>
            <Forms.Field
              inputRef={ref => (this._firstNameRef = ref)}
              autoFocus={true}
              onSubmitEditing={this.onSubmitEditing}
              field="firstName"
              label="First name"
              autoCorrect={false}
              style={styles.largerField}/>
          </View>
          <View style={Styles.Horizontal}>
            <Forms.Field
              inputRef={ref => (this._lastNameRef = ref)}
              onSubmitEditing={this.onSubmitEditing}
              field="lastName"
              label="Last name"
              autoCorrect={false}
              style={styles.largerField}/>
          </View>
          <View>
            <View style={Styles.Horizontal}>
              <Forms.Field
                inputRef={ref => (this._streetRef = ref)}
                onSubmitEditing={this.onSubmitEditing}
                field="street"
                label="Address"
                style={styles.largerField}/>
            </View>
            <View style={Styles.Horizontal}>
              <Forms.Field
                field="addressOpt"
                label="Optional Apt, Suite or Floor"
                style={styles.largerField}
                onSubmitEditing={this.onSubmitEditing}/>
            </View>
            <Forms.Field
              inputRef={ref => (this._cityRef = ref)}
              field="city"
              label="City"
              onSubmitEditing={this.onSubmitEditing}/>
            <View style={Styles.Horizontal}>
              <Forms.Field
                field="region"
                label="State/Province/Region"
                onSubmitEditing={this.onSubmitEditing}
                style={styles.largerField}/>
              <Forms.Field
                inputRef={ref => (this._postalRef = ref)}
                field="postal"
                label="Postal Code"
                onSubmitEditing={this.onSubmitEditing}/>
            </View>
            <Forms.Field
              inputRef={ref => (this._countryRef = ref)}
              onPress={this.pickCountry}
              onFocus={this.pickCountry}
              field="country"
              label="Country"/>
          </View>
          <View style={styles.buttons}>
            <Forms.Button onPress={() => this.submit()}>
              {this.submitButton}
            </Forms.Button>
            {this.shouldUpdate ? (
              <Forms.Button
                isEnabled
                onPress={() => this.delete()}
                color={Colours.DisabledButton}
                labelColor={Colours.Text}>
                {this.deleteButton}
              </Forms.Button>
            ) : null}
          </View>
        </ScrollView>
      </KeyboardAvoidingView>
    );
  };

  onPressManual = () => {
    this.setState({ manual: true });
  };

  render() {
    return (
      <Provider formStore={this.store}>
        <SafeAreaView style={styles.container}>
          {this.store.data.street || this.state.manual ? (
            this.renderManualForm()
          ) : (
            <GoogleAddress
              field={"address"}
              onPressManual={this.onPressManual}/>
          )}
        </SafeAreaView>
      </Provider>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colours.Foreground
  },

  content: {
    backgroundColor: Colours.Foreground,
    paddingTop: Sizes.OuterFrame,
    paddingBottom: Sizes.IOSTabBar + Sizes.OuterFrame * 2
  },

  largerField: {
    flex: 1
  },

  buttons: {
    marginTop: Sizes.InnerFrame
  }
});
