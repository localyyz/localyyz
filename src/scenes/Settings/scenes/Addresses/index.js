import React from "react";
import { View, StyleSheet, FlatList } from "react-native";

// third party
import { inject, observer } from "mobx-react/native";

// custom
import { Sizes } from "localyyz/constants";
import { BaseScene, Forms } from "localyyz/components";

@inject(stores => ({
  addresses: stores.addressStore.addresses,
  fetch: stores.addressStore.fetch
}))
@observer
export default class Addresses extends React.Component {
  constructor(props) {
    super(props);
    this.settings = this.props.navigation.state.params || {};

    // bindings
    this.renderItem = this.renderItem.bind(this);
    this.onPress = this.onPress.bind(this);
    this.onSelect = this.onSelect.bind(this);
    this.addAddress = this.addAddress.bind(this);
    this.updateAddress = this.updateAddress.bind(this);
    this.fetch = this.fetch.bind(this);
  }

  componentDidMount() {
    this.fetch();
  }

  async fetch() {
    let addresses = await this.props.fetch();

    // forward to add address by default if none exist
    if (!addresses || addresses.length < 1) {
      this.addAddress();
    }
  }

  renderItem({ item: address }) {
    return (
      <Forms.BaseField
        label={address.address}
        onPress={() => this.onPress(address)}/>
    );
  }

  onPress(address) {
    !this.onSelect(address) && this.updateAddress(address);
  }

  onSelect(address) {
    if (this.settings.onSelect) {
      this.settings.onSelect(address);

      // and out
      this.props.navigation.goBack(null);

      // help skip updateAddress on onPress
      return true;
    }
  }

  addAddress() {
    return this.props.navigation.navigate("AddressForm", {
      onSubmit: this.onSelect
    });
  }

  updateAddress(address) {
    return this.props.navigation.navigate("AddressForm", {
      shouldUpdate: true,
      address: address
    });
  }

  get footer() {
    return (
      <View style={styles.footer}>
        <Forms.Button isEnabled onPress={() => this.addAddress()}>
          Add a new address
        </Forms.Button>
      </View>
    );
  }

  render() {
    return (
      <BaseScene backAction={this.props.navigation.goBack} title="Addresses">
        <View style={styles.content}>
          <FlatList
            scrollEnabled={false}
            data={this.props.addresses && this.props.addresses.slice()}
            keyExtractor={address => `${address.id}`}
            renderItem={this.renderItem}
            ListFooterComponent={this.footer}/>
        </View>
      </BaseScene>
    );
  }
}

const styles = StyleSheet.create({
  content: {
    flex: 1
  },

  footer: {
    alignItems: "center",
    marginVertical: Sizes.InnerFrame,
    marginHorizontal: Sizes.OuterFrame
  }
});
