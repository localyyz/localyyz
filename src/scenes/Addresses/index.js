import React from "react";
import { HeaderBackButton, createStackNavigator } from "react-navigation";

// local
import Addresses from "./Addresses";
import Form from "./Form";
import CountryPicker from "./CountryPicker";

export default createStackNavigator(
  {
    Addresses: Addresses,
    AddressForm: Form,
    CountryPicker: CountryPicker
  },
  {
    navigationOptions: ({ navigation, navigationOptions }) => ({
      ...navigationOptions,
      headerLeft: (
        <HeaderBackButton
          tintColor="#000"
          onPress={() => navigation.goBack(null)}/>
      )
    })
  }
);
