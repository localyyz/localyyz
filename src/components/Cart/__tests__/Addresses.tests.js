import React from "react"
import Addresses from "../components/Addresses"
import Renderer from "react-test-renderer"

describe("Addresses", () => {

    it("Addresses: should render properly", () => {
        let props = {
            title: "Sample Address",
            addressStore : {
                addresses: [],
                fetch: () => {},
            }, 
            cartStore: {
                removeAddress: () => {},
                updateAddress: () => {}
            },
        }
        let rendered = Renderer.create(<Addresses {...props}/>).getInstance().wrappedInstance
        expect(rendered.refs.header.props.title).toBe("Sample Address")
        expect(rendered.refs.header.props.icon.props.name).toBe("dot-single")
        expect(rendered.refs.header.props.children).toBe("no address selected")
    })
})