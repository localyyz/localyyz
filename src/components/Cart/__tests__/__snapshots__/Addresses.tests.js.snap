// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressSelect should match snapshot when editing 1`] = `
<Component>
  <inject-AddressForm
    address={Object {}}
    isBillng={false}
    isShipping={false}
    onCancel={[MockFunction]}
    onSubmit={[MockFunction]}
    testID="addressForm"
  />
</Component>
`;

exports[`AddressSelect should match snapshot when selecting addresses 1`] = `
<Component>
  <Component>
    <Address
      address={
        UserAddress {
          "address": "123 Toronto street",
          "addressOpt": "",
          "city": "toronto",
          "country": "canada",
          "countryCode": undefined,
          "error": undefined,
          "firstName": "paul",
          "hasError": false,
          "id": 1,
          "isBilling": undefined,
          "isPartial": undefined,
          "isShipping": undefined,
          "lastName": "localyyz",
          "province": "ontario",
          "provinceCode": undefined,
          "zip": "abc 789",
        }
      }
      buttonColor="rgb(30, 30, 30)"
      onEdit={[MockFunction]}
      onPress={[Function]}
      onRemove={[Function]}
    />
    <Address
      address={
        UserAddress {
          "address": "180 john street",
          "addressOpt": "",
          "city": "toronto",
          "country": "canada",
          "countryCode": undefined,
          "error": undefined,
          "firstName": "paul2",
          "hasError": false,
          "id": 2,
          "isBilling": undefined,
          "isPartial": undefined,
          "isShipping": undefined,
          "lastName": "localyyz",
          "province": "ontario",
          "provinceCode": undefined,
          "zip": "xyz 123",
        }
      }
      buttonColor="rgb(30, 30, 30)"
      onEdit={[MockFunction]}
      onPress={[Function]}
      onRemove={[Function]}
    />
    <AddressNew
      enabled={true}
      onPress={[MockFunction]}
      testID="addressNew"
    />
  </Component>
</Component>
`;

exports[`Addresses should match snapshot when complete 1`] = `
<Component
  style={Object {}}
>
  <CartHeader
    icon={
      <Icon
        allowFontScaling={false}
        color="#9013FE"
        name="check-circle"
        size={10}
        style={
          Object {
            "marginRight": 6,
          }
        }
      />
    }
    title="Sample Address"
  >
    123 Toronto street
  </CartHeader>
  <AddressSelect
    addresses={
      Array [
        UserAddress {
          "address": "123 Toronto street",
          "addressOpt": "",
          "city": "toronto",
          "country": "canada",
          "countryCode": undefined,
          "error": undefined,
          "firstName": "paul",
          "hasError": false,
          "id": 1,
          "isBilling": undefined,
          "isPartial": undefined,
          "isShipping": undefined,
          "lastName": "localyyz",
          "province": "ontario",
          "provinceCode": undefined,
          "zip": "abc 789",
        },
        UserAddress {
          "address": "180 john street",
          "addressOpt": "",
          "city": "toronto",
          "country": "canada",
          "countryCode": undefined,
          "error": undefined,
          "firstName": "paul2",
          "hasError": false,
          "id": 2,
          "isBilling": undefined,
          "isPartial": undefined,
          "isShipping": undefined,
          "lastName": "localyyz",
          "province": "ontario",
          "provinceCode": undefined,
          "zip": "xyz 123",
        },
      ]
    }
    isBilling={false}
    isEditing={false}
    isShipping={false}
    onCancel={[Function]}
    onCreate={[Function]}
    onEdit={[Function]}
    onRemove={[Function]}
    onSelect={[Function]}
    selectedAddress={
      UserAddress {
        "address": "123 Toronto street",
        "addressOpt": "",
        "city": "toronto",
        "country": "canada",
        "countryCode": undefined,
        "error": undefined,
        "firstName": "paul",
        "hasError": false,
        "id": 1,
        "isBilling": undefined,
        "isPartial": undefined,
        "isShipping": undefined,
        "lastName": "localyyz",
        "province": "ontario",
        "provinceCode": undefined,
        "zip": "abc 789",
      }
    }
    testID="addressSelect"
  />
</Component>
`;

exports[`Addresses should match snapshot when incomplete 1`] = `
<Component
  style={Object {}}
>
  <CartHeader
    icon={
      <Icon
        allowFontScaling={false}
        color="#FF0000"
        name="dot-single"
        size={20}
      />
    }
    title="Sample Address"
  >
    no address selected
  </CartHeader>
  <AddressSelect
    addresses={
      Array [
        UserAddress {
          "address": "123 Toronto street",
          "addressOpt": "",
          "city": "toronto",
          "country": "canada",
          "countryCode": undefined,
          "error": undefined,
          "firstName": "paul",
          "hasError": false,
          "id": 1,
          "isBilling": undefined,
          "isPartial": undefined,
          "isShipping": undefined,
          "lastName": "localyyz",
          "province": "ontario",
          "provinceCode": undefined,
          "zip": "abc 789",
        },
        UserAddress {
          "address": "180 john street",
          "addressOpt": "",
          "city": "toronto",
          "country": "canada",
          "countryCode": undefined,
          "error": undefined,
          "firstName": "paul2",
          "hasError": false,
          "id": 2,
          "isBilling": undefined,
          "isPartial": undefined,
          "isShipping": undefined,
          "lastName": "localyyz",
          "province": "ontario",
          "provinceCode": undefined,
          "zip": "xyz 123",
        },
      ]
    }
    isBilling={false}
    isEditing={true}
    isShipping={false}
    onCancel={[Function]}
    onCreate={[Function]}
    onEdit={[Function]}
    onRemove={[Function]}
    onSelect={[Function]}
    selectedAddress={
      Object {
        "address": "123 Toronto street",
        "addressOpt": "",
        "city": "toronto",
        "country": "canada",
        "countryCode": undefined,
        "error": undefined,
        "firstName": "paul",
        "hasError": true,
        "id": 1,
        "isBilling": undefined,
        "isPartial": undefined,
        "isShipping": undefined,
        "lastName": "localyyz",
        "province": "ontario",
        "provinceCode": undefined,
        "zip": "abc 789",
      }
    }
    testID="addressSelect"
  />
</Component>
`;
